index/segment/mem/segment.go

```diff
@@ -93,24 +95,38 @@ func NewSegment(offset postings.ID, opts Options) (sgmt.MutableSegment, error) {
 
 func (s *segment) Insert(d doc.Document) error {
 	s.state.RLock()
+	defer s.state.RUnlock()
 	if s.state.closed {
-		s.state.RUnlock()
 		return sgmt.ErrClosed
 	}
 
 	if s.state.sealed {
-		s.state.RUnlock()
 		return errSegmentSealed
 	}
 
-	// Validate that the document contains only valid UTF-8.
-	for _, f := range d.Fields {
-		if !utf8.Valid(f.Name) {
-			return fmt.Errorf("document contains invalid field name: %v", f.Name)
+	docID, err := d.Validate()
+	if err != nil {
+		return err
+	}
+
+	if docID != nil {
+		ok, err := s.termsDict.ContainsTerm(doc.IDReservedFieldName, docID)
+		if err != nil {
+			return fmt.Errorf("could not lookup document ID '%s'", docID)
+		}
+		if ok {
+			// Index already contains this document.
+			return nil
 		}
-		if !utf8.Valid(f.Value) {
-			return fmt.Errorf("document contains invalid field value: %v", f.Value)
+	} else {
+		uuid, err := s.newUUIDFn()
```

Comment:
sorry should have been clearer, I meant optionally allow users to configure whether a uuid should be autogen'd or if this should be an error.

Reply:
oh, hmm, I can't really think of when that would be the case and if it is necessary I think it's fair to say it should be handled at a layer above the segment for now

Reply:
make an error function